selectel/mks.go

```diff
@@ -626,3 +628,122 @@ func parseMKSKubeVersionsV1Default(versions []*kubeversion.View) string {
 
 	return defaultVersion
 }
+
+// findQuota finds and returns quota for specified resource from quotas slice.
+func findQuota(quotas []*quotas.Quota, resource string) []quotas.ResourceQuotaEntity {
+	for _, q := range quotas {
+		if q.Name == resource {
+			return q.ResourceQuotasEntities
+		}
+	}
+
+	return nil
+}
+
+func checkQuotasForCluster(projectQuotas []*quotas.Quota, region string, zonal bool) error {
+	var (
+		clusterQuotaChecked bool
+		quota               []quotas.ResourceQuotaEntity
+	)
+
+	if zonal {
+		quota = findQuota(projectQuotas, "mks_cluster_zonal")
+	} else {
+		quota = findQuota(projectQuotas, "mks_cluster_regional")
+	}
+
+	if quota == nil {
+		return errors.New("unable to find neither mks_cluster_zonal nor mks_cluster_zonal quotas")
+	}
+
+	for _, v := range quota {
+		if v.Region == region {
+			if v.Value-v.Used <= 0 {
+				if zonal {
+					return errors.New("not enough quota to create zonal k8s cluster")
+				}
+
+				return errors.New("not enough quota to create regional k8s cluster")
+			}
+			clusterQuotaChecked = true
+		}
+	}
+	if !clusterQuotaChecked {
+		return errors.New("unable to check neither regional nor zonal k8s cluster quota for a given region")
+	}
+
+	return nil
+}
+
+func checkQuotasForNodegroup(projectQuotas []*quotas.Quota, nodegroupOpts *nodegroup.CreateOpts) error {
+	var cpuQuotaChecked, ramQuotaChecked, diskQuotaChecked bool
+
+	cpuQuota := findQuota(projectQuotas, "compute_cores")
+	if cpuQuota == nil {
+		return errors.New("unable to find CPU quota")
+	}
+	ramQuota := findQuota(projectQuotas, "compute_ram")
+	if ramQuota == nil {
+		return errors.New("unable to find RAM quota")
+	}
+
+	var volumeQuota []quotas.ResourceQuotaEntity
+	if nodegroupOpts.LocalVolume {
+		volumeQuota = findQuota(projectQuotas, "volume_gigabytes_local")
+	} else {
+		switch strings.Split(nodegroupOpts.VolumeType, ".")[0] {
+		case "fast":
+			volumeQuota = findQuota(projectQuotas, "volume_gigabytes_fast")
+		case "universal":
+			volumeQuota = findQuota(projectQuotas, "volume_gigabytes_universal")
+		case "basic":
+			volumeQuota = findQuota(projectQuotas, "volume_gigabytes_basic")
+		default:
+			return fmt.Errorf("expected 'fast.<zone>', 'universal.<zone>' or 'basic.<zone>' volume type, got: %s", nodegroupOpts.VolumeType)
+		}
+	}
+	if volumeQuota == nil {
+		return errors.New("unable to find volume quota")
+	}
+
+	requiredCPU := nodegroupOpts.CPUs * nodegroupOpts.Count
+	requiredRAM := nodegroupOpts.RAMMB * nodegroupOpts.Count
+	requiredVolume := nodegroupOpts.VolumeGB * nodegroupOpts.Count
+
+	for _, v := range cpuQuota {
+		if v.Zone == nodegroupOpts.AvailabilityZone {
+			if v.Value-v.Used < requiredCPU {
+				return fmt.Errorf("not enough CPU quota to create nodegroup, free: %d, required: %d", v.Value-v.Used, requiredCPU)
```

Comment:
When we update existing nodegroup this error message is confusing.
We have to rewrite it to have the same meaning for the create and update cases.